### PR TITLE
nixos/sshd: Specify connection parameters for configuration validation

### PR DESCRIPTION
**Pull request by CRTified**
## Description of changes

After #253427, I had [the problem that my `Match`rules trigger a failing config verification](https://github.com/NixOS/nixpkgs/pull/253427#issuecomment-1724408816). I did not create an issue so far.

This PR 
 - Tests with specific `laddr`,`lport` as described in the documentation for the `-C` flag (see below). This iterates over all specified `ports` and `listenAddresses` from the nix configuration.
 - Adds a test case that has all possible `Match` rules (at the time of writing) to test for this type of configurations.

Note that the test would fail without the change to the config validation.

One remaining thing to discuss would be whether there is a way of testing the other connection parameters:
```
       -C connection_spec
              Specify the connection parameters to use for the -T extended test mode.  If provided, any Match direc‐
              tives  in  the  configuration file that would apply are applied before the configuration is written to
              standard output.  The connection parameters are supplied as keyword=value pairs and may be supplied in
              any order, either with multiple -C options or as a comma-separated list.  The keywords  are  ‘‘addr'',
              ‘‘user'',  ‘‘host'', ‘‘laddr'', ‘‘lport'', and ‘‘rdomain'' and correspond to source address, user, re‐
              solved source host name, local address, local port number and routing domain respectively.
```

Relevant pings: @Ma27 @Yarny0 

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Ma27** [comment]
Thanks a lot for providing a fix already! Have seen your comment on the other PR, but haven't had the time to take a look so far, sorry!

Will try to take a look soonish.

**Comment by CRTified** [comment]
No problem, I think that there is a small discussion required on how to handle the other parameters.

It's also not entirely clear to me why we need the `-T` flag here, but without it, `sshd` won't take the parameters supplied with `-C`, not solving the problem.

**Comment by CRTified** [comment]
So, to summarize: This PR is fine as is, but could be improved by two things:

The first thing are the remaining `Match` criterias. I've just tested *all* criterias and apparently it's only `LocalAddress` and `LocalPort` that are sensitive in the check. All other values are also highly dynamic (user, remote address and so on), so I could imagine (but I did **not** verify this with upstream source code) that they are actually treated differently.

I'll update the test case (i.e., add a "large" `extraConfig` section) and force-push it in a minute, feedback on whether this suffices is highly desired.


The other thing to optimize is parallelization/off-loading the validation into a separate derivation. I'll check it out and report back.

**Comment by CRTified** [comment]
So, I've moved the validation to `system.checks`. If I remember `nativeBuildInputs` vs. `buildInputs` correctly, that should also cross-"compile" correctly (using native tooling on the builder for config validation).

**Comment by CRTified** [comment]
After noticing the `validationPackage` thing, I finally think that it's now in a state to discuss (or even merge). Sorry for the noise.


Only thing I'm unhappy with is the fact that we now create a config from `cfg.settings` and only then merge the `extraConfig` in a separate `runCommand`: https://github.com/NixOS/nixpkgs/pull/256090/files#diff-909cff3303f72af4beab9cc703252041321d4df4d5fe94729505a35165787fa4R30

It would be great if we could make `Match` rules expressible with the generation from `cfg.settings` (making `extraConfig` obsolete), but that is probably out of scope for this PR.

**Comment by CRTified** [comment]
I've dropped the `exit 1` now. I'd assume that it should fail, too, but I simply looked at other instances for `system.checks` usage and found it in use there.

**Review by Ma27 — APPROVED** [review]
Thanks a lot for your work :+1: 
AFAIU this is not only a bugfix for the regression I caused with my merge (sorry again :sweat_smile: ), but even an improvement since multiple Match clauses being actually applied are also tested now.

I'd give some more folks a chance to review, but I'd merge tomorrow otherwise.

**Comment by happysalada** [comment]
hey, I keep on getting
```
error: builder for '/nix/store/c6pbpw5hjkjgipmarwyic9zyqr1xaix5-check-sshd-config.drv' failed with exit code 1;
       last 1 log lines:
       > Invalid test mode specification -f
       For full logs, run 'nix log /nix/store/c6pbpw5hjkjgipmarwyic9zyqr1xaix5-check-sshd-config.drv'
```

it seems related to this PR.
here is my ssh config.
```
{
  enable = true;
  controlMaster = "auto";
  controlPath = "~/.ssh/control/%r@%h:%p";
  controlPersist = "5m";
  compression = true;
  matchBlocks = {
    "bee" = {
      hostname = "142.170.43.99";
      # hostname = "192.168.2.10";
      user = "yt";
    };
    "hetz" = {
      hostname = "116.202.222.51";
      user = "yt";
    };
  };
  matchBlocks."*" = {
    extraOptions = {
      Protocol = "2";

      # also this stuff
      TCPKeepAlive = "yes";
      ServerAliveInterval = "20";
      ServerAliveCountMax = "10";

      # stop bugging me about a new host
      StrictHostKeyChecking = "no";

      # run gui programs on local
      ForwardX11 = "yes";
      UseRoaming = "no";
      KexAlgorithms =
        "curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256";
      HostKeyAlgorithms =
        "ssh-ed25519-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,ssh-ed25519,ssh-rsa";
      Ciphers =
        "chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr";
      PubkeyAuthentication = "yes";
      MACs =
        "hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,umac-128@openssh.com";
      PasswordAuthentication = "no";
      ChallengeResponseAuthentication = "no";
      # UseKeychain = "yes";
      AddKeysToAgent = "yes";
      identityFile = "~/.ssh/id_ed25519";
      ForwardAgent = "yes"; # enables use of local ssh agent (e.g. for github)
    };
  };
}
```

maybe this is my fault, I'm happy to change whatever is needed.

**Comment by CRTified** [comment]
@happysalada The error indicates that it is caused by this PR, yes.

But this PR changed the `sshd` config validation (i.e., the server side). Can you share your config for that?

I'm generally surprised here that the error appears to be caused by the `-f`.

**Comment by Ma27** [comment]
@happysalada this is your home-manager config which is not relevant to this change at all. A minimal Nix expression (perhaps even in the form of a VM test) and a nixpkgs git revision where you got this error would be greatly appreciated.

**Comment by happysalada** [comment]
this is my ssh config for the server (my bad for the first time).
```
  # agent can only be started as non root
  programs.ssh.startAgent = true;

  services = {
    openssh = {
      enable = true;
      settings.PermitRootLogin = "prohibit-password";
      settings.KbdInteractiveAuthentication = false;
      settings.PasswordAuthentication = false;
      listenAddresses = [{ addr = "0.0.0.0"; port = 22;} {addr = "[::]"; port = 22;}];
    };
  };
```

nixpkgs is on latest unstable at rev f3dab3509afca932f3f4fd0908957709bb1c1f57

**Comment by CRTified** [comment]
Could be caused by the IPv6 address (and the lack of escaping). Let me check, it might be a simple fix.

**Comment by Ma27** [comment]
@crtified Yeah, it's an escaping problem. Using `escapeShellArg` solves the issue for me. Currently preparing a fix.

**Comment by Ma27** [comment]
@CRTified @happysalada #257721 solves the problem for me. Also modified the openssh test to make sure it'd fail without the fix.

File: nixos/modules/services/networking/ssh/sshd.nix [hunk 1]
```diff
@@ -576,6 +574,21 @@ in
...
+        } ''
+        ${concatMapStringsSep "\n"
+          (lport: "sshd -G -T -C lport=${toString lport} -f ${sshconf} > /dev/null")
+          cfg.ports}
+        ${concatMapStringsSep "\n"
+          (la: "sshd -G -T -C laddr=${la.addr},lport=${toString la.port} -f ${sshconf} > /dev/null")
+          cfg.listenAddresses}
+        touch $out
+      '')
+    ];
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ssh/sshd.nix:590]: A comment what we are checking here wouldn't have hurt.

Comment (Ma27) [reply nixos/modules/services/networking/ssh/sshd.nix:590]: The drv name is pretty descriptive though, isn't it?

**Pull request merged**